c/src/atom.rs

```diff
@@ -396,6 +412,26 @@ pub unsafe extern "C" fn atom_get_children(atom: *const atom_t,
     }
 }
 
+/// Performs a depth-first exhaustive iteration of an atom and all its children recursively.
+/// The first result returned will be the atom itself
+#[no_mangle]
+pub unsafe extern "C" fn atom_iterate(atom: *const atom_t,
+        callback: c_atom_callback_t, context: *mut c_void) {
+    let atom = &(*atom).atom;
+    for inner_atom in AtomIter::new(atom) {
```

Comment:
The short answer is because the prototype of `Space::query` in Rust is: `fn query(&self, query: &Atom) -> BindingsSet;`, and thus the C wrapper function gets called with `const atom_t*`.

Of course a simple cast can easily sweep away such concerns in C, and a mutable iterate function is more useful.  The Rust programmer in me wants to expose both a mutable and immutable version of the iterate function, but it seems a little bit pointless since the implementations are basically identical.

I will make atom_iterate() mutable, and just do the cast.  :-P  When in Rome...

Reply:
Yes I also felt introducing both mutable and immutable in a C layer is overhead. This is why I asked the question. But looking at the mutable implementation I feel we could have both :-) Anyway each version has only three lines of code, but using `*mut` can make an additional unnecessary confusion in a C implementation. 

I am sorry for this misleading question above Luke, don't you mind removing https://github.com/trueagi-io/hyperon-experimental/commit/ed4307bfc9618c0062ed59d5dba8a998d57a4a7a from the branch, and I will merge it?